File: backend/app/services/fare_calculator.py
"""
Indian Railways Telescopic Fare Calculation Service.

Computes passenger fare estimates based on authentic distance slabs,
reservation charges, and superfast surcharges across coach classes.
Labelled transparently as CALCULATED_ESTIMATE.
"""
from __future__ import annotations

import math
from typing import Dict, Any, Optional

# Authentic Indian Railways Telescopic Base Fare Coefficients & Surcharges
CLASS_FARE_CONFIG = {
    "SL": {
        "name": "Sleeper Class",
        "base_rate_per_km": 0.48,
        "reservation_fee": 20,
        "superfast_surcharge": 20,
        "min_distance_km": 200,
        "min_fare": 140,
    },
    "CC": {
        "name": "AC Chair Car",
        "base_rate_per_km": 1.15,
        "reservation_fee": 40,
        "superfast_surcharge": 45,
        "min_distance_km": 150,
        "min_fare": 385,
    },
    "3A": {
        "name": "AC 3 Tier",
        "base_rate_per_km": 1.35,
        "reservation_fee": 40,
        "superfast_surcharge": 45,
        "min_distance_km": 300,
        "min_fare": 495,
    },
    "2A": {
        "name": "AC 2 Tier",
        "base_rate_per_km": 1.95,
        "reservation_fee": 50,
        "superfast_surcharge": 45,
        "min_distance_km": 300,
        "min_fare": 710,
    },
    "1A": {
        "name": "AC First Class",
        "base_rate_per_km": 3.30,
        "reservation_fee": 60,
        "superfast_surcharge": 75,
        "min_distance_km": 300,
        "min_fare": 1250,
    },
    "EC": {
        "name": "Executive Chair Car",
        "base_rate_per_km": 2.10,
        "reservation_fee": 60,
        "superfast_surcharge": 75,
        "min_distance_km": 150,
        "min_fare": 835,
    },
}


def compute_haversine_distance_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculate great-circle distance between two GPS coordinates in kilometers."""
    R = 6371.0  # Earth radius in kilometers
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(dlon / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return R * c


def estimate_class_fare(coach_class: str, distance_km: float) -> int:
    """
    Compute estimated one-way ticket fare for a given class and distance.
    Rounds to nearest 5 rupees (Indian Railways standard rounding rule).
    """
    norm_class = coach_class.upper().strip()
    config = CLASS_FARE_CONFIG.get(norm_class)
    if not config:
        # Fallback to general 3A default estimation
        config = CLASS_FARE_CONFIG["3A"]

    dist = max(distance_km, float(config["min_distance_km"]))
    # Telescopic slab discount for very long journeys (> 500km and > 1000km)
    if dist > 1000:
        telescopic_factor = 0.85
    elif dist > 500:
        telescopic_factor = 0.92
    else:
        telescopic_factor = 1.0

    raw_fare = (dist * config["base_rate_per_km"] * telescopic_factor) + config["reservation_fee"] + config["superfast_surcharge"]
    fare = max(int(raw_fare), config["min_fare"])

    # Round to nearest 5 rupees
    rounded_fare = int(round(fare / 5.0) * 5)
    return rounded_fare


def calculate_segment_fares(
    from_lat: Optional[float],
    from_lon: Optional[float],
    to_lat: Optional[float],
    to_lon: Optional[float],
    track_distance_km: Optional[float] = None,
) -> Dict[str, Any]:
    """
    Calculate full class fare matrix for a station-to-station journey segment.
    """
    if track_distance_km and track_distance_km > 10:
        dist_km = track_distance_km
    elif from_lat and from_lon and to_lat and to_lon:
        # Railway tracks have approx ~1.25 winding factor compared to straight line
        straight_km = compute_haversine_distance_km(from_lat, from_lon, to_lat, to_lon)
        dist_km = max(straight_km * 1.25, 50.0)
    else:
        dist_km = 300.0  # Sensible default trunk distance

    dist_km = round(dist_km, 1)

    fares_by_class = {
        cls: estimate_class_fare(cls, dist_km)
        for cls in ["SL", "3A", "2A", "1A", "CC", "EC"]
    }

    return {
        "distance_km": dist_km,
        "fares": fares_by_class,
        "min_fare": min(fares_by_class.values()),
        "fare_type": "CALCULATED_ESTIMATE",
        "fare_source_label": "Estimated Fare (IR Telescopic Tariff)",
        "fare_disclaimer": "Tariffs calculated via Indian Railways standard telescopic distance slabs. Dynamic flexi-fares or Tatkal surcharges may apply at PRS charting.",
    }

File: frontend/src/utils/fareCalculator.ts
/**
 * Client-Side Indian Railways Telescopic Fare Calculator.
 *
 * Implements standard distance-based telescopic fare slabs, reservation fees,
 * and superfast surcharges across coach classes (1A, 2A, 3A, SL, CC, EC).
 * Fares are transparently labeled as CALCULATED_ESTIMATE.
 */

export interface ClassFareEstimate {
  coachClass: string;
  name: string;
  fare: number;
  fareType: 'CALCULATED_ESTIMATE';
  sourceLabel: string;
  disclaimer: string;
}

export interface SegmentFareSummary {
  distanceKm: number;
  minFare: number;
  fares: Record<string, number>;
  classBreakdown: ClassFareEstimate[];
  fareType: 'CALCULATED_ESTIMATE';
  sourceLabel: string;
  disclaimer: string;
}

const CLASS_RATES: Record<
  string,
  {
    name: string;
    ratePerKm: number;
    reservationFee: number;
    superfastFee: number;
    minDist: number;
    minFare: number;
  }
> = {
  SL: {
    name: 'Sleeper Class (SL)',
    ratePerKm: 0.48,
    reservationFee: 20,
    superfastFee: 20,
    minDist: 200,
    minFare: 140,
  },
  CC: {
    name: 'AC Chair Car (CC)',
    ratePerKm: 1.15,
    reservationFee: 40,
    superfastFee: 45,
    minDist: 150,
    minFare: 385,
  },
  '3A': {
    name: 'AC 3 Tier (3A)',
    ratePerKm: 1.35,
    reservationFee: 40,
    superfastFee: 45,
    minDist: 300,
    minFare: 495,
  },
  '2A': {
    name: 'AC 2 Tier (2A)',
    ratePerKm: 1.95,
    reservationFee: 50,
    superfastFee: 45,
    minDist: 300,
    minFare: 710,
  },
  '1A': {
    name: 'AC First Class (1A)',
    ratePerKm: 3.3,
    reservationFee: 60,
    superfastFee: 75,
    minDist: 300,
    minFare: 1250,
  },
  EC: {
    name: 'Executive Chair Car (EC)',
    ratePerKm: 2.1,
    reservationFee: 60,
    superfastFee: 75,
    minDist: 150,
    minFare: 835,
  },
};

/**
 * Calculate Haversine distance between two coordinates in kilometers.
 */
export function calculateDistanceKm(
  lat1?: number,
  lon1?: number,
  lat2?: number,
  lon2?: number
): number {
  if (!lat1 || !lon1 || !lat2 || !lon2) return 300; // Default trunk distance
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  // Multiply by track curvature factor (~1.25)
  return Math.max(Math.round(R * c * 1.25), 45);
}

/**
 * Estimate fare for a single coach class given journey distance in km.
 */
export function estimateClassFare(coachClass: string, distanceKm: number): number {
  const norm = coachClass.toUpperCase().trim();
  const cfg = CLASS_RATES[norm] || CLASS_RATES['3A'];
  const dist = Math.max(distanceKm, cfg.minDist);

  let telescopicFactor = 1.0;
  if (dist > 1000) telescopicFactor = 0.85;
  else if (dist > 500) telescopicFactor = 0.92;

  const raw = dist * cfg.ratePerKm * telescopicFactor + cfg.reservationFee + cfg.superfastFee;
  const fare = Math.max(Math.round(raw), cfg.minFare);
  // Round to nearest 5 rupees (IR standard)
  return Math.round(fare / 5) * 5;
}

export function calculateFare(distanceKm: number, coachClass: string): { fare: number; sourceLabel: string } {
  return {
    fare: estimateClassFare(coachClass, distanceKm),
    sourceLabel: 'IR Telescopic Tariff',
  };
}

/**
 * Calculate full class fare matrix for a station pair.
 */
export function calculateSegmentFareSummary(
  fromLat?: number,
  fromLon?: number,
  toLat?: number,
  toLon?: number,
  knownDistanceKm?: number
): SegmentFareSummary {
  const distanceKm = knownDistanceKm || calculateDistanceKm(fromLat, fromLon, toLat, toLon);
  const classes = ['SL', '3A', '2A', '1A', 'CC', 'EC'];

  const fares: Record<string, number> = {};
  const classBreakdown: ClassFareEstimate[] = [];

  for (const cls of classes) {
    const f = estimateClassFare(cls, distanceKm);
    fares[cls] = f;
    classBreakdown.push({
      coachClass: cls,
      name: CLASS_RATES[cls]?.name || cls,
      fare: f,
      fareType: 'CALCULATED_ESTIMATE',
      sourceLabel: 'Estimated Fare',
      disclaimer: 'Based on Indian Railways standard telescopic distance slabs.',
    });
  }

  const minFare = Math.min(...Object.values(fares));

  return {
    distanceKm,
    minFare,
    fares,
    classBreakdown,
    fareType: 'CALCULATED_ESTIMATE',
    sourceLabel: 'Estimated Fare (IR Telescopic Tariff)',
    disclaimer:
      'Fares computed from Indian Railways distance tariff slabs. Dynamic flexi-fares or Tatkal surcharges may apply at PRS charting.',
  };
}
